### PR TITLE
feat: Generate wrapper gem for cloud

### PR DESCRIPTION
**Pull request by dazuma**
This PR builds on a previous refactor where the presenters were moved out of the `templates` directory into the `lib` directory (#371). That refactor makes it easier to subclass the presenters.

The major changes:

* Created directory `gapic-generator-cloud/templates/cloud/wrapper_gem` with a bunch of new templates specific to creating the wrapper gem.
* Created presenter subclasses specific to the wrapper gem in `gapic-generator-cloud/lib/gapic/presenters/*.rb`. These wrapper-specific presenters override some methods in the base presenters, and add others specific to the wrapper templates.
* Modified `gapic-generator-cloud/lib/gapic/generators/cloud_generator.rb` to run a different generation script (the `generate_wrapper` method) if a certain configuration is provided. The new script instantiates the wrapper presenter subclasses. Then, it generates some files from the base templates and others from the wrapper-specific templates.
* Several additional configuration keys were added to control content and behavior of wrapper generation. You can see these in the changes to `gapic-generator-cloud/bin/ruby-cloud-docker-entrypoint`. Methods were added to the presenters that read these values and interpret them for the templates.

Minor and incidental changes:

* When generating a wrapper, we do not want to invoke the `ruby` or `grpc` proto plugins, so we update the protoc command invocations accordingly. This appears in `gapic-generator-cloud/Rakefile` as well as in the Docker entrypoint `gapic-generator-cloud/bin/ruby-cloud-docker-entrypoint`.
* The template that generates AUTHENTICATION.md has been updated to work for either a gapic library or a wrapper library. (The differences are in the sample code, and are controlled by calls to the presenters. The base presenter returns text for the gapic case, whereas the wrapper presenter subclasses return text for the wrapper case.)
* The template that generates LICENSE.md was not well-formatted markdown. I updated it to format a reasonable markdown file. This affects the gapic libraries in addition to the wrapper libraries.
* Enhanced the doc formatter to allow disabling generation of YARD cross-reference links. This allows us to include the service docs in the YARD for the wrapper's client factory methods. Links in those docs need to be disabled because the classes they would link to are classes in the versioned gapic client, not the wrapper.
* Standardized the use of the article "the" and the "API" qualifier, in service titles (e.g. "_the_ Natural Language _API_") that appear in READMEs, gemspecs, and YARD. This affects gapic libraries in addition to wrapper libraries.
* Removed some redundant vertical whitespace in many generated files that use `gapic-generator/templates/default/layouts/_ruby.erb`. This affects the gapic libraries in addition to the wrapper libraries.

Also added the secret_manager wrapper to the outputs, to demonstrate the results of the wrapper generator.

Some sample PRs showing the difference between the wrapper generator and current handwritten wrappers:

* google-cloud-secret_manager (a new library) : https://github.com/googleapis/google-cloud-ruby/pull/5092
* google-cloud-language (a migrated library) : https://github.com/googleapis/google-cloud-ruby/pull/5093

### CODE REVIEW

**Comment by dazuma** [comment]
I've now opened PRs using this generator to update _all_ existing wrapper libraries in google-cloud-ruby. Feel free to refer to those PRs to understand the effect of the wrapper generator.

File: gapic-generator-cloud/templates/cloud/wrapper_gem/_main.erb [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+require "<%= gem.name %>"
+<%- end -%>
+<% end %>
+<%- gem.services.each do |service| -%>
+##
+# Create a new client object for <%= service.name %>.
+#
+# By default, this returns an instance of
+# <%= gem.docs_link version: gem.default_version, class_name: "#{service.name}::Client" %>
+# for version <%= gem.default_version.capitalize %> of the API.
```
Comment (viacheslav-rostovtsev) [inline gapic-generator-cloud/templates/cloud/wrapper_gem/_main.erb]: feels like should be a sub-template

File: gapic-generator-cloud/templates/cloud/wrapper_gem/main.erb [hunk 2]
```diff
@@ -0,0 +1,5 @@
+<%- assert_locals gem -%>
```
Comment (viacheslav-rostovtsev) [inline gapic-generator-cloud/templates/cloud/wrapper_gem/main.erb]: why does this redirect-template exist?

File: gapic-generator/lib/gapic/presenters/gem_presenter.rb [hunk 3]
```diff
@@ -121,6 +121,34 @@ def iam_dependency?
...
+      def yard_strict?
+        gem_config(:yard_strict) ? true : false
+      end
+
+      def entrypoint_require
+        packages.first.package_require
+      end
+
+      def needs_dashed_ruby_file?
+        name != namespace_file_path
```
Comment (viacheslav-rostovtsev) [inline gapic-generator/lib/gapic/presenters/gem_presenter.rb]: where does this come from? seems to be defined in the subclass of this class

File: gapic-generator/lib/gapic/presenters/gem_presenter.rb [hunk 4]
```diff
@@ -121,6 +121,34 @@ def iam_dependency?
...
+      def product_documentation_url
+        gem_config :product_documentation_url
+      end
+
+      def api_id
+        gem_config :api_id
+      end
+
+      def free_tier?
+        gem_config(:free_tier) ? true : false
```
Comment (viacheslav-rostovtsev) [inline gapic-generator/lib/gapic/presenters/gem_presenter.rb]: seems like this and some other fields are used only in wrapper templates but are defined here

File: gapic-generator-cloud/templates/cloud/wrapper_gem/readme.erb [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+   to get started.
+<%- else -%>
+1. [Enable billing for your project.](https://cloud.google.com/billing/docs/how-to/modify-project#enable_billing_for_a_project)
+<%- end -%>
+<%- if gem.api_id -%>
+1. [Enable the API.](https://console.cloud.google.com/apis/library/<%= gem.api_id %>.googleapis.com)
+<%- end -%>
+1. {file:AUTHENTICATION.md Set up authentication.}
+
+<%- if gem.migration? %>
```
Comment (viacheslav-rostovtsev) [inline gapic-generator-cloud/templates/cloud/wrapper_gem/readme.erb]: Optional for the greenfield libraries? if pre_migration_version is nil perhaps?

File: gapic-generator-cloud/templates/cloud/wrapper_gem/main.erb (hunk 2 above)
Comment (dazuma) [reply gapic-generator-cloud/templates/cloud/wrapper_gem/main.erb]: The ruby layout requires its content to be a partial. (You can see this structure in the default gapic-generator templates also, `service/client.erb` and others.)

File: gapic-generator/lib/gapic/presenters/gem_presenter.rb (hunk 3 above)
Comment (dazuma) [reply gapic-generator/lib/gapic/presenters/gem_presenter.rb]: Oh, that's a good catch. The `needs_dashed_ruby_file?` method should have been defined in the subclass.

File: gapic-generator/lib/gapic/presenters/gem_presenter.rb (hunk 4 above)
Comment (dazuma) [reply gapic-generator/lib/gapic/presenters/gem_presenter.rb]: This one will be used in the main generator. It's used in the README, and I want to do a follow-up for the main generator that includes a lot of this content there as well.

File: gapic-generator-cloud/templates/cloud/wrapper_gem/readme.erb (hunk 5 above)
Comment (dazuma) [reply gapic-generator-cloud/templates/cloud/wrapper_gem/readme.erb]: This is exactly what `gem.migration?` tests: whether the library is being migrated as opposed to greenfield. (And yes, its implementation is essentially, is the migration version nil or not.)

File: gapic-generator-cloud/templates/cloud/wrapper_gem/_main.erb (hunk 1 above)
Comment (dazuma) [reply gapic-generator-cloud/templates/cloud/wrapper_gem/_main.erb]: Generally we've been using partials for places where we think there's a reasonable chance that a particular section will want to be overridden by a subgenerator. Partials are the "unit" over overriding templates. In this case, I don't expect that we will have a subgenerator under the wrapper generator. If we ever do, we can easily refactor to break out a partial at that time.

**Pull request merged**